### ISSUE TITLE
Option to disable E1029 using a regex pattern on the variable name

### ISSUE THREAD

**Issue by coyoteecd** [aws-cloudformation/cfn-python-lint#1278]
*cfn-lint version: 0.26.0*

This is another take on the https://github.com/aws-cloudformation/cfn-python-lint/issues/740. I am using cfn-lint in combination with Serverless Framework and the VSCode plugin to validate the CloudFormation template "chunks" that are referenced in the resource section within the serverless.yml (spec: https://serverless.com/framework/docs/providers/aws/guide/serverless.yml/)

`serverless.yml` (**not** validated by cfn-lint, provided for reference)

```
service: my-service
provider:
  name: aws
[... other stuff ..]

# importing raw CloudFormation files here
resources:
  - Description: My Example Service Stack
  # roles
  - ${file(./cfn-resources/resource1.yml)}
  - ${file(./cfn-resources/resource2.yml)}
```

`.cfnlintrc`
```
templates:
- cfn-resources/*.yml
```

`resource1.yml`
```
Conditions:
  PrimaryRegion: !Equals ['${self:provider.region}', 'us-east-1']

[some resources]
```

Serverless framework will pre-process the yml file and replace the ${..} sequence with actual string values before generating the actual CloudFormation template to be sent to AWS. Unfortunately they chose a syntax that conflicts with something CloudFormation uses as well.

With the above config, cfn-lint generates this error:

`E1029: Found an embedded parameter outside of an "Fn::Sub" at Conditions/PrimaryRegion/Fn::Equals/0`

I can get around this error by putting a !Sub in front of it (which is useless, it will upload an unnecessary !Sub 'us-east-1' to AWS). But then I get this:

`E1019: Parameter self:provider.region for Fn::Sub not found at Conditions/PrimaryRegion/Fn::Equals/0/Fn::Sub`

So I have to disable both rules, or disable E1019 and stick a !Sub in front of everything that has a serverless-specific variable.

It would be great to have a way to be able to configure both rules in such a way that they treat the whole `${variable}`  block as a literal by if `variable` matches a regex pattern. Ideally through .cfnlintrc so it applies to all the files...

NOTE: I am aware that this is not an issue with cfn-python-lint, but I don't see how it could be solved by the VSCode plugin (it basically reports what it gets from cfn-lint) or the Serverless team (they're stuck with the syntax due to backwards compatibility, it's been discussed before, in the context of having the problem in the opposite direction by serverless treating ${AWS::Region} as its own variable and trying to replace it, see https://github.com/serverless/serverless/pull/3694)

**Comment by kddejong** [comment]
I think this could be done easily.  Configuring this rule seems to be never ending so making it a little more dynamic seems like a reasonable approach to me.

**Comment by coyoteecd** [comment]
@kddejong assuming you're hinting at a pull request: I haven't written any python yet, but if you thinking it's easy I can give it a shot when I have time.

For confirmation, I would assume the change involves something like:
- choose a way to configure the offending rule (E1019), for example:
```
configure_rules:
  E1019:
    resource_excludes_custom_filter: [some regex]
    condition_excludes_custom_filter: [other regex]
```
- use self.config to read the rule settings in [match_values function around here](https://github.com/aws-cloudformation/cfn-python-lint/blob/d1e58cbcba8bb2ef9c14badc98c0ebe70fe4cbbf/src/cfnlint/rules/functions/SubNeeded.py#L106-L115) and decide whether to skip the rule checks in a similar way as already done with the hard-coded resource_excludes and condition_excludes
- also add a self.config_definition in the same class, in a similar way to [this one](https://github.com/aws-cloudformation/cfn-python-lint/blob/b788cc9bd3d49ed20d5f2e58602755a0ef37f52c/src/cfnlint/rules/resources/properties/ValuePrimitiveType.py#L26-L31)
- update docs, mentioning the settings in https://github.com/aws-cloudformation/cfn-python-lint#configure-rules
- write a couple of tests

If I missed something, let me know.

**Comment by kddejong** [comment]
@coyoteecd I don't mind trying to tackle this too.  It looks like you are heading in the right direction.  Let me know what works best for you.

**Comment by coyoteecd** [comment]
After a bit of thought, I realized that the initial config proposal is not what I need; the reason is the resource/condition excludes are meant to apply specifically to the Resource/NotResource/Condition subsections, while in my case I have "variables" that must be ignored in other places too, e.g. name of a resource.

So I settled for fixing E1029 to have a generic custom exclude regex:
```
configure_rules:
  E1029:
    custom_excludes: [some regex]
```

The regex matches against the complete variable name, i.e. `${self:provider.stage}`, so an example regex would be:
```yaml
configure_rules:
  E1029:
    custom_excludes: ^\${self.+}$
```
The above ignores any variable that starts with "self" anywhere in the .yml file.

**Comment by coyoteecd** [comment]
I installed 0.28.1 and it works as intended, we can close this.
